modules/terrain/src/lib/parse-npy.js

```diff
@@ -0,0 +1,98 @@
+/* globals TextDecoder */
+
+function systemIsLittleEndian() {
+  const a = new Uint32Array([0x12345678]);
+  const b = new Uint8Array(a.buffer, a.byteOffset, a.byteLength);
+  return !(b[0] === 0x12);
+}
+
+const LITTLE_ENDIAN_OS = systemIsLittleEndian();
+
+// The basic string format consists of 3 characters:
+// 1. a character describing the byteorder of the data (<: little-endian, >: big-endian, |: not-relevant)
+// 2. a character code giving the basic type of the array
+// 3. an integer providing the number of bytes the type uses.
+// https://numpy.org/doc/stable/reference/arrays.interface.html
+//
+// Here I only include the second and third characters, and check endianness
+// separately
+const DTYPES = {
+  u1: Uint8Array,
+  i1: Int8Array,
+  u2: Uint16Array,
+  i2: Int16Array,
+  u4: Uint32Array,
+  i4: Int32Array,
+  f4: Float32Array,
+  f8: Float64Array
+};
+
+export function parseNPY(arrayBuffer, options) {
+  if (!arrayBuffer) {
+    return null;
+  }
+
+  const view = new DataView(arrayBuffer);
+  const {header, headerEndOffset} = parseHeader(view);
+
+  const numpyType = header.descr;
+  const ArrayType = DTYPES[numpyType.slice(1, 3)];
+  if (!ArrayType) {
+    // eslint-disable-next-line no-console, no-undef
+    console.warn(`Decoding of npy dtype not implemented: ${numpyType}`);
+    return null;
+  }
+
+  const nArrayElements = header.shape.reduce((a, b) => a * b);
+  const arrayByteLength = nArrayElements * ArrayType.BYTES_PER_ELEMENT;
+
+  const data = new ArrayType(arrayBuffer.slice(headerEndOffset, headerEndOffset + arrayByteLength));
+
+  // Swap endianness if needed
+  if ((numpyType[0] === '>' && LITTLE_ENDIAN_OS) || (numpyType[0] === '<' && !LITTLE_ENDIAN_OS)) {
+    // eslint-disable-next-line no-console, no-undef
+    console.warn('Data is wrong endianness, byte swapping not yet implemented.');
```

Comment:
*Years late*, but was just passing by and saw this. I did this the silly way recently before realizing that the [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) interface makes this exceptionally easy and probably [has good enough support](https://caniuse.com/?search=dataview) to rely upon these days.

Reply:
Is this a use case for you? You're trying to load big-endian npy data?

Reply:
No, not directly. I was just working through this to get a handle on npy but this jumped out since I just worked through parsing COG headers and took a nice little detour into bit shifts and array views (which don't work because the data may not be aligned) before realizing DataView is a pretty good interface for this. And with apologies, I saw after writing this that you've exactly used DataView just below! Pardon the distraction! 🤦

Reply:
yeah dataview is a great API! I just didn't implement this here because I figured big-endian data was such an edge case hah. Also calling dataview in a loop for every item in an array wouldn't be that performant